전남대 BE_강명덕 3주차 과제 (0단계)

Profile-exe opened the pull request:
저번주 코드를 옮기고 나서 추가로 리팩토링과 테스트코드를 작성했습니다.

## 구현 내용
- `member`, `product` 도메인의 리파지토리 예외처리 코드를 서비스로 옮겼습니다.
- `wish list`, `product`의 E2E 테스트 코드를 작성했습니다. Step1을 구현해야하는데 테스트코드를 작성하느라 시간이 많이 지체되어 `member` 도메인 E2E 테스트는 추후 작성할 예정입니다.

## 코드를 작성하며 어려웠던 점
- `TestRestTemplate`를 처음 접해 테스트코드를 짜는데 어려움이 있었습니다.
- 토큰 기반 인증을 구현하다보니 테스트케이스별로 인증을 한 상태에서 요청을 보내야했습니다. 이 부분을 구현하는게 어려웠습니다.

## 코드 리뷰 시, 멘토님이 중점적으로 리뷰해줬으면 하는 부분
- 인증을 한 상태로 요청을 보내기 위해 `@BeforeAll`을 통해 한 계정만 로그인한 후 테스트를 지속해서 진행하려했으나 `static` 메서드만 해당 어노테이션 사용이 가능했습니다. 그래서 `@Autowired`를 이용한 `TestRestTemplate` 의존성 주입이 불가능했습니다. 

  다른 방식을 찾아보다 `@BeforeEach`를 사용해 각 테스트 시점에서 서로 다른 이메일로 가입을 한 후 해당 토큰으로 테스트케이스 로직을 수행하도록 작성했습니다. 이러한 방식이 괜찮은지, `@BeforeAll`을 사용하면서도 `static` 변수로 바뀌게 될  `TestRestTemplate`을 주입받아 이용할 수 있는 방법이 있는지 궁금합니다.

Profile-exe committed:
init: 2주차 위시 리스트 코드 반영

Profile-exe committed:
refactor: repository에서 진행하던 예외처리를 service로 이동

회원, 상품 도메인도 예외처리를 서비스에서 하도록 변경

Profile-exe committed:
feat(wish_list): 외래키에 cascade 제약조건 추가

이를 통해 member, product 테이블 데이터 삭제 가능

Profile-exe committed:
feat(TestUtils): ResponseEntity 생성용 유틸 클래스 구현

Profile-exe committed:
feat(ErrorResponse): 역직렬화를 위한 생성자 구현

Profile-exe committed:
test(ProductControllerTest): 상품 API E2E 테스트 구현

Profile-exe committed:
feat(WishListController): 장바구니 추가 API 응답 개선

- ResponseEntity.ok() 대신 ResponseEntity.created()를 사용하여 201 상태 코드 반환
- URI.create()를 통해 새로 생성된 리소스의 위치 지정
- body()를 사용하여 응답 메시지 포함

Profile-exe committed:
feat(WishListService): 존재하는 상품에 대해서만 위시리스트에 넣을 수 있도록 검증 추가

Profile-exe committed:
test(WishListControllerTest): 위시 리스트 API E2E 테스트 구현

Profile-exe committed:
delete(EndToEnd): 불필요한 코드 제거

hocaron commented:
이러한 방식이 괜찮은지, @BeforeAll을 사용하면서도 static 변수로 바뀌게 될 TestRestTemplate을 주입받아 이용할 수 있는 방법이 있는지 궁금합니다.
테스트 코드 시나리오에 대한 좋은 고민이네요! 저는 명덕님께서 말씀 주신 방법도 너무 좋은 것 같고, 이 외에 2가지 방법이 떠오르는데요.

1. Lifecycle.PER_CLASS 를 통해 @BeforeAll 내부 메서드를 non-static 을 테스트하기
fyi; https://www.baeldung.com/java-beforeall-afterall-non-static

2. @AfterEach 에 데이터 클렌징 로직 추가하여 테스트 실행 후에 데이터를 초기화하거나 제거하는 방법

저는 지금의 방법도 좋은 것 같아요. 그 다음 요청에서 중복되는 데이터가 생기는 케이스를 없애는 방법도 좋습니다. 명덕님은 3가지 방법에 대해 어떻게 생각하시나요?!😀

hocaron reviewed (approved):
명덕님, 안녕하세요👋

완벽한 E2E 테스트네요👍
1단계 진행을 위해 테스트 코드 관련 논의가 끝나면 바로 머지할게요! 이번 3주차도 화이팅입니다💪

Profile-exe committed:
test(ProductControllerTest): 회원과는 연관없는 도메인이므로 단일 계정 테스트로 처리

Profile-exe committed:
test(WithListControllerTest): 테스트 케이스마다 가입 후 위시 리스트 초기화 및 탈퇴 적용

테스트케이스마다 서로다른 여러 회원을 가입시키지 않고 하나의 회원에 대해

 가입 - 위시 리스트 등록 - 테스트 진행 - 탈퇴

 순으로 진행되도록 수정

Profile-exe commented:
질문에 대해 답변해주신 부분

> 1. Lifecycle.PER_CLASS 를 통해 @BeforeAll 내부 메서드를 non-static 을 테스트하기
>    fyi; https://www.baeldung.com/java-beforeall-afterall-non-static
> 2. @AfterEach 에 데이터 클렌징 로직 추가하여 테스트 실행 후에 데이터를 초기화하거나 제거하는 방법
> 
> 저는 지금의 방법도 좋은 것 같아요. 그 다음 요청에서 중복되는 데이터가 생기는 케이스를 없애는 방법도 좋습니다. 명덕님은 3가지 방법에 대해 어떻게 생각하시나요?!😀

`Lifecycle.PER_CLASS`라는 유용한 기능이 있었네요! 감사합니다.

<br>

#### 방법 1
`@BeforeAll`사용은 상품 도메인에 적절하다 생각했습니다. 상품 도메인은 회원과 연관관계가 없기 때문에 테스트를 위한 인터셉터의 인가 허용만 필요합니다. 따라서 인가 허용을 위한 회원가입을 테스트 전에 단 한번만 진행한 후 테스트를 하면 된다고 생각합니다.
  
  이 부분은 `ProductControllerTest`에 적용했으며 [0d174772](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-jpa/pull/70/commits/0d174772d1ec584177fc700aa40eb7335a9ae996) 커밋에 반영했습니다.

<br>

#### 방법2
`@AfterEach`에 데이터 초기화나 제거 방법은 위시 리스트가 적절하다 생각했습니다. `@BeforeEach`에서 회원가입을 진행한 후 위시 리스트 테스트에 사용할 상품 3개를 장바구니에 담고 `@AfterEach`에서 테스트가 끝난 계정을 삭제하는 식으로 진행했습니다. 

여기서 `1) 기존의 위시 리스트 내역 삭제`와 `2) 가입한 회원 삭제` 두 가지 방법이 있습니다. 저는 `2) 가입한 회원 삭제` 방법을 적용했는데, 위시 리스트 테이블 명세는 [b8c50aed](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-jpa/pull/70/commits/b8c50aedb43e824a7e1d6c3c632b4b181405f8d6) 커밋에서 `cascade` 제약조건을 적용했습니다. 따라서 회원 탈퇴로 위시 리스트 데이터도 함께 지워지며 코드 로직상으로도 `회원 가입 -> 테스트 진행 -> 회원 탈퇴`가 자연스럽기 때문에 방법 `2)`를 적용했습니다.

그리고 회원 탈퇴는 요청 1번이면 가능하지만, 위시 리스트를 비우는  방식은 위시 리스트 개수만큼 진행하므로 API 호출이 그만큼 많아져 비효율적이라 생각했습니다. '내 장바구니 비우기' API등을 구현할 수 있지만 현재 상황에서 굳이 필요하지 않다고 생각했습니다.

이 부분은 `WishListControllerTest`에 적용했으며 [a3164e53](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-jpa/pull/70/commits/a3164e53cdbdc60b5838ee22117c87a3148a5c33) 커밋에 반영했습니다.

<br>

#### 방법3
요청에서 중복되는 데이터가 생기는 케이스를 없애는 방법은 제가 처음 적용한 서로 다른 이메일 계정으로 가입하는 방식인 것 같은데, 서로 다른 이메일로 가입하게되면 테스트케이스 수에 따라 가입되는 계정 개수가 계속 늘어나며, `@AfterEach` 등으로 2번처럼 회원 탈퇴 처리한다면 이는 동일한 계정으로 가입-탈퇴를 진행하는 것과 동일하다고 보았습니다. 따라서 1,2번 방법을 각각 제가 적절하다 생각하는 테스트에 적용해보았습니다.
